bofire/strategies/samplers/universal_constraint.py

```diff
@@ -0,0 +1,58 @@
+import numpy as np
+import pandas as pd
+
+from bofire.data_models.domain.api import Domain
+from bofire.data_models.strategies.api import UniversalConstraintSampler as DataModel
+from bofire.strategies.doe.design import find_local_max_ipopt
+from bofire.strategies.enum import OptimalityCriterionEnum
+from bofire.strategies.strategy import Strategy
+
+
+class UniversalConstraintSampler(Strategy):
+    """Sampler that generates samples by optimization in IPOPT.
+
+    Attributes:
+        domain (Domain): Domain defining the constrained input space
+        sampling_fraction (float, optional): Fraction of sampled points to total points generated in
+            the sampling process. Defaults to 0.3.
+        ipopt_options (dict, optional): Dictionary containing options for the IPOPT solver. Defaults to {"maxiter":200, "disp"=0}.
+    """
+
+    def __init__(
+        self,
+        data_model: DataModel,
+        **kwargs,
+    ):
+        super().__init__(data_model=data_model, **kwargs)
+        assert data_model.sampling_fraction > 0 and data_model.sampling_fraction <= 1
+        self.sampling_fraction = data_model.sampling_fraction
+        self.ipopt_options = data_model.ipopt_options
+
+    def _ask(self, candidate_count: int) -> pd.DataFrame:
+        samples = find_local_max_ipopt(
```

Comment:
We could also include pending candidates here, or? Meaning that we provide already done experiments to the sampler to which the distance should be maximized? Those would be available via `self.candidates` via the `fixed_experiments` keyword of `find_local_max_ipopt`. What do you think?

Reply:
I assumed candidates means we do not necessarily need to have them in the set of samples (so we do not sample pending candidates w/ higher priority. Is this right?)

Reply:
Hi Aaron,

regarding the pending candidates, the idea is the following:

We provide a set of pending candindates, and want to sample `n` new candidates which are optimal given the objective and the set of pending ones.  The `find_local_max_ipopt` returns also the fixed ones if you provide them. So you have to increase the number of samples to get from `find_local_max_ipopt` and remove then the pending ones from the dataframe before you do the random sampling.

Do you see what I mean? Can you then also add a test where pending ones are used?

Best,

Johannes

Reply:
Hi Johannes, 

thx for the clarification! I think in the new commit it is done the right way.

Cheers,
Aaron :)